lib/diff/hex/hex.ex

```diff
@@ -69,29 +69,38 @@ defmodule Diff.Hex do
   end
 
   def diff(package, from, to) do
-    path_from = tmp_path("#{package}-#{from}-")
-    path_to = tmp_path("#{package}-#{to}-")
+    path_from = tmp_path("package-#{package}-#{from}-")
+    path_to = tmp_path("package-#{package}-#{to}-")
+    path_diff = tmp_path("diff-#{package}-#{from}-#{to}-")
 
     try do
       with {:ok, tarball_from} <- get_tarball(package, from),
-           {:ok, tarball_to} <- get_tarball(package, to),
            :ok <- unpack_tarball(tarball_from, path_from),
+           {:ok, tarball_to} <- get_tarball(package, to),
            :ok <- unpack_tarball(tarball_to, path_to),
-           {:ok, gd} <- git_diff(path_from, path_to),
-           {:ok, parsed} <- GitDiff.parse_patch(gd) do
-        {:ok, parsed}
+           :ok <- git_diff(path_from, path_to, path_diff) do
+        stream =
+          File.stream!(path_diff, [:read_ahead])
+          |> GitDiff.stream_patch(relative_from: path_from, relative_to: path_to)
+          |> Stream.transform(
+            fn -> :ok end,
+            fn elem, :ok -> {[elem], :ok} end,
+            fn :ok -> File.rm(path_diff) end
+          )
+
+        {:ok, stream}
       else
         error ->
-          Logger.error("Failed to diff #{package} #{from}..#{to} with: #{inspect(error)}")
-          {:error, :unknown}
+          Logger.error("Failed to create diff #{package} #{from}..#{to} with: #{inspect(error)}")
+          :error
       end
     after
       File.rm_rf(path_from)
       File.rm_rf(path_to)
     end
```

Comment:
Should we attempt to remove `path_diff` here too, in case something raised?

Reply:
We remove it after the stream is done on line https://github.com/hexpm/diff/pull/46/files#diff-0098d6d39b513f644bf5c7f4de6481adR88 because the stream needs to read from it. There's a chance we leak it if we never start the stream, but since we store it in the temp directory it should be eventually cleaned I think.

Reply:
Think it's only cleared on reboots, but I guess those will happen often enough to prevent build up

Reply:
Okay, worst case kubernetes will reboot if we run out of disk space =).